### PR TITLE
add documentation to explain how to deploy cilium on k0s

### PR DESCRIPTION
**Pull request by xinity**
Please ensure your pull request adheres to the following guidelines:

- [X] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [X] Thanks for contributing!

this documentation explain how to deploy cilium on k0s using the k0sctl tooling.

```release-note
docs: Add install guide for deploying Cilium on K0s using k0sctl.
```

### CODE REVIEW

File: Documentation/installation/k0s.rst
Comment (qmonnet) [inline Documentation/installation/k0s.rst]:
```suggestion
*****************************
Installation k0s Using k0sctl
*****************************
```

File: Documentation/installation/k0s.rst
Comment (qmonnet) [inline Documentation/installation/k0s.rst]: Nothing particular against the smiley, but I don't think we're usually using them in the docs so it feels a bit strange to me to have one here.

```suggestion
The first step (after installing the k0sctl binary obviously) is to prepare the yaml configuration file k0sctl will use:
```

File: Documentation/installation/k0s.rst
Comment (qmonnet) [inline Documentation/installation/k0s.rst]:
```suggestion
   # The following command assumes the user has deployed 3 VMs
   # with the default user "k0s" using the default ssh-key (without passphrase)
```

File: Documentation/installation/k0s.rst
Comment (qmonnet) [inline Documentation/installation/k0s.rst]: This is no shell code. Just a literal block.

```suggestion
Next step is editing ``k0s-myk0scluster-config.yaml``::
```

File: Documentation/installation/k0s.rst
Comment (qmonnet) [inline Documentation/installation/k0s.rst]:
```suggestion
    # edit :k0s-myk0scluster-config.yaml: 
```

File: Documentation/installation/k0s.rst
Comment (qmonnet) [inline Documentation/installation/k0s.rst]:
```suggestion
Next step is editing ``k0s-myk0scluster-config.yaml``:
```

File: Documentation/installation/k0s.rst
Comment (qmonnet) [inline Documentation/installation/k0s.rst]:
```suggestion
    # replace
```

File: Documentation/installation/k0s.rst
Comment (qmonnet) [inline Documentation/installation/k0s.rst:87]:
```suggestion
generate the ``kubeconfig`` file, store it in ``~/.kube/k0s-mycluster.config`` and set
```

**Review by qmonnet — CHANGES_REQUESTED** [review]
Thanks a lot! Some formatting elements to fix, please see inline below.

I'd merge the two commits into a single one, it's better to have the docs build correctly on each commit (it won't build if the doc is added but not linked anywhere).

One last thing, would you please consider using a real email address in the `Signed-off-by:` tag for the commit?

File: Documentation/installation/k0s.rst
Comment (qmonnet) [inline Documentation/installation/k0s.rst]: CI workflow also caught this: You have to update the reference target above.

**Comment by qmonnet** [comment]
In addition to the formatting fixes, I got the following feedback on the PR:

- It would be great if we had a CI job to test Cilium on k0s, to make sure that the steps from the guide remain valid and that Cilium still installs correctly in the future. Something like [the one for Kind](.github/workflows/conformance-kind.yaml), possibly. Do you think you could work on that?

- If you are using Cilium, would you please consider adding yourself (or your company, if relevant) to the file `USERS.md`? We try to gather a list of users, in part to help with the CNCF graduation for the project.

**Comment by xinity** [comment]
@qmonnet related to your questions:
> In addition to the formatting fixes, I got the following feedback on the PR:

> It would be great if we had a CI job to test Cilium on k0s, to make sure that the steps from the guide remain valid and that Cilium still installs correctly in the future. Something like [the one for Kind](https://github.com/cilium/cilium/pull/.github/workflows/conformance-kind.yaml), possibly. Do you think you could work on that?

==> i will look at it sure :)

> If you are using Cilium, would you please consider adding yourself (or your company, if relevant) to the file USERS.md? We try to gather a list of users, in part to help with the CNCF graduation for the project.

==> will open a PR right after this one is done :)

File: Documentation/installation/k0s.rst
Comment (qmonnet) [inline Documentation/installation/k0s.rst]:
```suggestion
    # edit k0s-myk0scluster-config.yaml:  
```

Sorry, my mistake on the previous suggestion. No colon needed here.

File: Documentation/installation/k0s.rst
Comment (qmonnet) [inline Documentation/installation/k0s.rst]:
```suggestion
The first step (after installing the k0sctl binary obviously) is to prepare the yaml configuration file k0sctl will use:
```

File: Documentation/installation/k0s.rst
Comment (qmonnet) [inline Documentation/installation/k0s.rst]: My bad on the first pass

```suggestion
   # The following command assumes the user has deployed 3 VMs
```

File: Documentation/installation/k0s.rst
Comment (qmonnet) [inline Documentation/installation/k0s.rst]: I think we need to remove the space if we want the colon to show up here. But that's not really important anyway.

```suggestion
Next step is editing ``k0s-myk0scluster-config.yaml``::
```

**Review by qmonnet — APPROVED** [review]
Looks good to me, thanks again!

Again, I would encourage you to switch to a real email in the `Signed-off-by:` tag in the commit description. We've occasionally had to reach out to contributors in the past.

Not marking as ready-to-merge just yet, I'm trying to find someone who knows more than I do about k0s to take a look at the guide. Please ping if you get no news within a few days.

**Comment by xinity** [comment]
conformance test added @qmonnet :)

File: .github/workflows/conformance-k0s.yaml
Comment (joestringer) [inline .github/workflows/conformance-k0s.yaml]: What's the purpose of the SSH configuration?

File: Documentation/installation/k0s.rst
Comment (joestringer) [inline Documentation/installation/k0s.rst]: Please explicitly request the user to install k0sctl, something like:

```suggestion
Ensure you have the k0sctl binary installed locally.

Prepare the yaml configuration file k0sctl will use:
```

The first sentence should give a link to the install docs upstream. It could be placed above the "Install a Master Node" title.

File: .github/workflows/conformance-k0s.yaml
Comment (joestringer) [reply .github/workflows/conformance-k0s.yaml]: Hmm, I see, this is basically reconfiguring the github worker to become the one k8s node in the cluster (via the k0s-config.yaml).

File: Documentation/installation/k0s.rst
Comment (joestringer) [inline Documentation/installation/k0s.rst:39]: This seems like something that should be an explicit step, something like "Set up your VMs. How to do this is out of scope of this guide. After deploying the VMs, take note of the IP addresses of the VMs on the local network. These will be used in a later step.".

You could even get the reader to define an environment variable with these here, then later on in the k0sctl command just replace the current hard-coded IP addresses with the variable.

**Review by joestringer — CHANGES_REQUESTED** [review]
I have a couple of pieces of feedback on the documentation steps. For the github workflows, I'd like a second opinion from @cilium/github-sec .

File: .github/workflows/conformance-k0s.yaml
Comment (joestringer) [reply .github/workflows/conformance-k0s.yaml]: ^ Github visually displays a red bar for the tab indentation here too, I suspect it expects space-indentation. While we're at it, this should probably be changed. I don't know if that impacts the ability of github to parse the workflow.

**Comment by joestringer** [comment]
I didn't try this out, but the instructions seem mostly straightforward. I actually wonder whether we really need to run a conformance test for this platform given how little is different for it, this seems like a fairly standard k8s distribution with no unusual instructions for the interaction between Cilium and that distribution.

File: .github/workflows/conformance-k0s.yaml
Comment (xinity) [reply .github/workflows/conformance-k0s.yaml]: that's exactly it :) no worries at all :)

**Comment by xinity** [comment]
@joestringer updates implemented :)
thanks a lot for the feedback :)

File: .github/workflows/conformance-k0s.yaml
Comment (joestringer) [inline .github/workflows/conformance-k0s.yaml]: There's still a couple of places where it looks like this github workflow is misformatted. If you visit https://github.com/cilium/cilium/pull/22029/files, you can see how github highlights those portions, highlighted in a dark colour.

File: .github/workflows/conformance-k0s.yaml
Comment (joestringer) [inline .github/workflows/conformance-k0s.yaml]: @cilium/github-sec, @cilium/ci-structure  What do you think about the frequency of running this platform test?

My impression is that there is almost no k0s-specific configuration here, so the existing kind checks are likely to catch most regressions before we release a version. That makes me inclined to think that we shouldn't run this workflow for every PR change or merge, but instead either run it once as part of release processes (maybe after branching), or skip the workflow entirely and rely on the k0s community to perform this testing and report any issues observed.

Thoughts?

File: Documentation/installation/k0s.rst
Comment (joestringer) [inline Documentation/installation/k0s.rst:17]: This statement about platform support is generally true, but having this statement here in these docs makes it more likely that we will forget to update this if we change the set of supported platforms. Should amd64/arm64 platform support be mentioned on some more central page, maybe https://docs.cilium.io/en/latest/operations/system_requirements/ ?

File: Documentation/installation/k0s.rst
Comment (joestringer) [inline Documentation/installation/k0s.rst]: I think that the instructions to disable kube-proxy should perhaps be indented further to match the indentation aligmnent of the "If running ..." sentence, so that they show up inside the note box. You should be able to test this out and ensure it shows up inside that box by using `make render-docs` locally.

**Review by joestringer — COMMENTED** [review]
Just a couple more minor nits, and otherwise we need to figure out amongst @cilium/github-sec and @cilium/ci-structure how this platform should be integrated into automated testing.

File: .github/workflows/conformance-k0s.yaml
Comment (christarazi) [reply .github/workflows/conformance-k0s.yaml]: I would agree, assuming that this workflow will have some flakiness eventually. More tests == more potential flakes, especially in the case where we already have coverage, I don't see what value running this would bring.

Comment (xinity) [reply .github/workflows/conformance-k0s.yaml]: @christarazi @joestringer should i simply drop the k0s conformance test then ?

Comment (joestringer) [reply .github/workflows/conformance-k0s.yaml]: I posted this on Slack to gather more perspectives but I'm not seeing anyone arguing to keep this workflow. So I think yeah maybe we can just delete the conformance test. It's nice to have a way to confirm it works, but the docs should be sufficient to allow us to validate this. Sorry to go back and forth with you on this, but yes please delete it.

Comment (nbusseneau) [reply .github/workflows/conformance-k0s.yaml]: I'm late to the party, but my opinion is that CI is OK as long as we have someone in the community maintaining the CI. If we do not, then when CI breaks it'll just get disabled because of lack of cycles to maintain.

FWIW, we already have install guides dedicated to specific platforms / distributions (basically half of the things listed in https://docs.cilium.io/en/latest/installation/k8s-toc/, which is where k0s currently intends to go, and everything from this page: https://docs.cilium.io/en/latest/installation/external-toc/), there is no CI for these and it seems to be mostly OK to me.

So my vote is on no CI.

**Comment by xinity** [comment]
should be all find now @joestringer @qmonnet @nbusseneau nbusseneau 

let me know :)

**Review by joestringer — APPROVED** [review]
LGTM, thanks!

**Comment by joestringer** [comment]
The BPF checks (likely just patch commit message checks) and Documentation updates tests are failing, please take a look.

**Review by nbusseneau — APPROVED** [review]

**Comment by xinity** [comment]
@joestringer should be all fixed by now 🤞

**Review by nbusseneau — CHANGES_REQUESTED** [review]
@xinity The documentation build is still failing due to suspected spelling errors: https://github.com/cilium/cilium/actions/runs/3748172185/jobs/6368682084

I think the `environnement` one is genuine and should be fixed. For `sctl` I'm actually not quite sure why it's detected since it's part of `k0sctl` (@qmonnet might know), but in any case you can run `Documentation/update-spelling_wordlist.sh environnement sctl` as suggested in the logs, so that `sctl` is added to the ignore list, and then amend your commit :)

**Comment by xinity** [comment]
seen that using the test-doc feature , i'm on it :)
will update ASAP :)

File: Documentation/installation/k0s.rst
Comment (nbusseneau) [inline Documentation/installation/k0s.rst]: Sorry if I wasn't clear in my previous message: `environnement` is a genuine spelling mistake, it should be `environment` instead. So this should be fixed + removed from `Documentation/spelling_wordlist.txt` ;)

**Review by nbusseneau — CHANGES_REQUESTED** [review]

File: Documentation/installation/k0s.rst
Comment (xinity) [reply Documentation/installation/k0s.rst]: haven't seen this mistake, will fix ASAP

**Review by nbusseneau — APPROVED** [review]

**Comment by nbusseneau** [comment]
The various failures in CI are due to the rebase on top of `master` after #21744 was merged. A fix will be merged independently. In any case, this PR only changes documentation, documentation CI passed and all reviews are in, marking ready to merge.

**Pull request merged**
